Fix JITless Wasm to JS stack parameter handling.

danlliu opened the pull request:
#### 5d23096ae0a41ec4503ef776e1a4a93ac7ce51db
<pre>
Fix JITless Wasm to JS stack parameter handling.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278188">https://bugs.webkit.org/show_bug.cgi?id=278188</a>
<a href="https://rdar.apple.com/133985209">rdar://133985209</a>

Reviewed by Yusuke Suzuki.

Previously, the Wasm to JS operation used isStackArgument. However, since the calling convention is reversed (Wasm is the caller), it is no longer a stack argument, and is just a stack location, meaning that we need to use isStack.

* JSTests/wasm/stress/cc-int-to-int-memory.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.let.wat2.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.table.t.1.funcref.import.string_appeared_here.string_appeared_here.func.test.param.x.i32.result.i32.import.string_appeared_here.string_appeared_here.func.callee.param.x.i32.result.i32.memory.1.data.i32.const.0.string_appeared_here.func.test2.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.test3.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.add.i32.add.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.const.5.call.callee.i32.load.i32.const.0.i32.add.i32.add.i32.add.i32.add.async test.callee):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.let.wat2.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.table.t.1.funcref.import.string_appeared_here.string_appeared_here.func.test.param.x.i32.result.i32.import.string_appeared_here.string_appeared_here.func.callee.param.x.i32.result.i32.memory.1.data.i32.const.0.string_appeared_here.func.test2.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.test3.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.add.i32.add.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.const.5.call.callee.i32.load.i32.const.0.i32.add.i32.add.i32.add.i32.add.async test):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.async test): Deleted.
* JSTests/wasm/stress/cc-parameters-overwhelming-to-js.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.func.callee.param.a.i32.param.b.i32.param.c.i32.param.d.i32.param.e.i32.param.f.i32.param.g.i32.param.h.i32.param.i.i32.param.j.i32.param.k.i32.param.l.i32.param.m.i32.param.n.i32.param.o.i32.param.p.i32.result.i32.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.const.0.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.i32.const.13.i32.const.14.i32.const.15.call.callee.callee):
(async test):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):

Canonical link: <a href="https://commits.webkit.org/282429@main">https://commits.webkit.org/282429@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4b800f7b3dc09e7391dc5007908a77366b9bd57

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15594 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67019 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13602 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50041 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13886 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50776 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9385 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66067 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39349 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54561 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31460 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36038 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11897 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12478 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/56108 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57582 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12228 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68714 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/62241 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6944 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11847 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58088 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6976 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54624 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58293 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13992 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5791 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/84004 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38174 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14783 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40365 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38996 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

justinmichaud commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -360,7 +360,7 @@ JSC_DEFINE_JIT_OPERATION(operationWasmToJSExitMarshalArguments, EncodedJSValue,
         case TypeKind::Externref:
         case TypeKind::Funcref:
         case TypeKind::I32: {
-            if (wasmParam.isStackArgument()) {
```
I can't leave a comment on the line, but:
```
wasmCallingConvention().callInformationFor(typeDefinition, CallRole::Callee);
```
You could also change that to caller, right? I wonder which makes more sense here

danlliu replied:
Agh, that's a good catch, that would probably be better.

danlliu committed:
Fix JITless Wasm to JS stack parameter handling.
https://bugs.webkit.org/show_bug.cgi?id=278188
rdar://133985209

Reviewed by Yusuke Suzuki.

Previously, the Wasm to JS operation used isStackArgument. However, since the calling convention is reversed (Wasm is the caller), it is no longer a stack argument, and is just a stack location, meaning that we need to use isStack.

* JSTests/wasm/stress/cc-int-to-int-memory.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.let.wat2.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.table.t.1.funcref.import.string_appeared_here.string_appeared_here.func.test.param.x.i32.result.i32.import.string_appeared_here.string_appeared_here.func.callee.param.x.i32.result.i32.memory.1.data.i32.const.0.string_appeared_here.func.test2.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.test3.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.add.i32.add.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.const.5.call.callee.i32.load.i32.const.0.i32.add.i32.add.i32.add.i32.add.async test.callee):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.let.wat2.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.table.t.1.funcref.import.string_appeared_here.string_appeared_here.func.test.param.x.i32.result.i32.import.string_appeared_here.string_appeared_here.func.callee.param.x.i32.result.i32.memory.1.data.i32.const.0.string_appeared_here.func.test2.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.test3.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.add.i32.add.func.export.string_appeared_here.param.x.i32.result.i32.i32.load.i32.const.0.call.test.i32.const.1.i32.load.i32.const.0.i32.const.5.call.callee.i32.load.i32.const.0.i32.add.i32.add.i32.add.i32.add.async test):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.memory.1.data.i32.const.0.string_appeared_here.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.load.i32.const.0.async test): Deleted.
* JSTests/wasm/stress/cc-parameters-overwhelming-to-js.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.import.string_appeared_here.string_appeared_here.func.callee.param.a.i32.param.b.i32.param.c.i32.param.d.i32.param.e.i32.param.f.i32.param.g.i32.param.h.i32.param.i.i32.param.j.i32.param.k.i32.param.l.i32.param.m.i32.param.n.i32.param.o.i32.param.p.i32.result.i32.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.const.0.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.i32.const.13.i32.const.14.i32.const.15.call.callee.callee):
(async test):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):

Canonical link: https://commits.webkit.org/282429@main
